tornettools/generate.py

```diff
@@ -220,26 +233,66 @@ def __server(args, network, server):
     # tgen starts at the end of shadow's "bootstrap" phase
     process["start_time"] = BOOTSTRAP_LENGTH_SECONDS
 
+    if server['is_hs_server']:
+        # this is an onion service, so tgen should only listen on localhost
+        process["environment"] = "TGENIP=127.0.0.1"
+
     host["processes"].append(process)
 
+    if server['is_hs_server']:
+        # prepare the hostname and hs_ed25519_secret_key files for the onion service
+        hosts_prefix = "{}/{}/{}".format(args.prefix, SHADOW_TEMPLATE_PATH, SHADOW_HOSTS_PATH)
+        server_prefix = "{}/{}".format(hosts_prefix, server['name'])
+        hs_prefix = "{}/{}".format(server_prefix, TOR_ONIONSERVICE_DIR)
+
+        if not os.path.exists(hs_prefix):
+            os.makedirs(hs_prefix, 0o700)
+
+        with open("{}/{}".format(hs_prefix, 'hostname'), 'w') as outf:
+            outf.write(server['hs_hostname']+'\n')
+        with open("{}/{}".format(hs_prefix, 'hs_ed25519_secret_key'), 'wb') as outf:
+            outf.write(b"== ed25519v1-secret: type0 ==\x00\x00\x00" + base64.b64decode(server['hs_ed25519_secret_key']))
+
+        # tor process for the onion service
+        process = {}
+        process["path"] = "{}/bin/tor".format(SHADOW_INSTALL_PREFIX)
+        process["args"] = __format_tor_args(None)
+        process["start_time"] = BOOTSTRAP_LENGTH_SECONDS-60 # start before boostrapping ends
+
+        host["processes"].append(process)
+
     return {server['name']: host}
 
 def __perfclient(args, network, client):
+    # a perfclient can have one of two tgen configurations which specifies which servers it connects to
+    if not client['is_hs_client']:
+        tgenrc_fname = TGENRC_PERFCLIENT_PUBLIC_FILENAME
+    else:
+        tgenrc_fname = TGENRC_PERFCLIENT_HS_FILENAME
+
     return __tgen_client(args, network, client['name'], client['country_code'], \
-        get_host_rel_conf_path(TGENRC_PERFCLIENT_FILENAME))
+        get_host_rel_conf_path(tgenrc_fname))
 
 def __markovclient(args, network, client):
     # these should be relative paths
     return __tgen_client(args, network, client['name'], client['country_code'], \
         TGENRC_MARKOVCLIENT_FILENAME)
 
 def __format_tor_args(name):
-    args = [
-        f"--Address {name}",
-        f"--Nickname {name}",
-        f"--defaults-torrc {TORRC_DEFAULTS_HOST_FILENAME}",
-        f"-f {TORRC_HOST_FILENAME}",
-    ]
+    args = []
+
+    if name is not None:
+        args.append(f"--Address {name}")
+
+        # tor nicknames must be between 1 and 19 characters inclusive
+        if len(name) < 1 or len(name) > 19:
+            logging.warning(f'Nickname {name} is too long and will likely cause tor to exit')
```

Comment:
Do we also want to give a hint to the user about how they can fix the issue?

Reply:
I see this more of a developer warning since the user doesn't really have control over the hostnames/nicknames, so I'm not sure how the user would fix this.

Reply:
OK. I suppose the only way a user would see this is if they generated big enough Tor networks to cause a name like
`relay012345exitguard`, but perhaps we don't have to worry about a million relays - yet, that is :P